### PR TITLE
Short article on UBSan trap-on-error

### PR DESCRIPTION
**Pull request by noahp**

### CODE REVIEW

File: _drafts/ubsan_trap.md [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+foo is
+```
+
+We get a weird result! we've indexed into the `arr` array at an out-of-bounds
+position (2), but the program happily fetched that memory location and tried to
+print it as a string 😬.
+
+This type of bug is very popular over at https://cve.mitre.org/ 😔
+
+#### WIth UBsan
```
Comment (dauc) [inline _drafts/ubsan_trap.md:113]: Typo, `WIth` (capital I).

File: _drafts/ubsan_trap.md [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+If we run that program with problematic input, we get an error, and our program
+crashes instead of doing something unexpected!
+
+![img/ubsan-trap/ubsan-basic.png](/img/ubsan-trap/ubsan-basic.png)
+
+## `libubsan` library
+
+For hosted platforms (eg Linux, BSD's, macOS, Windows), there's usually a
+`libusban` library installed with your toolchain. On my system it's here:
```
Comment (shivarajagopal) [inline _drafts/ubsan_trap.md:147]: Typo: `libubsan`

Comment (noahp) [reply _drafts/ubsan_trap.md:159]: :fire: USB :fire:

Comment (noahp) [reply _drafts/ubsan_trap.md]: heh i had 2 counts of USB in this article :sweat_smile: bless u sir

**Review by franc0is — APPROVED** [review]
This is excellent. Would be curious to see what running UBsan on e.g. one of the ChibiOS examples gets us.

**Review by chrisc11 — APPROVED** [review]
<3

**Comment by noahp** [comment]
> This is excellent. Would be curious to see what running UBsan on e.g. one of the ChibiOS examples gets us.

It was pretty informative on a small project I tested it on recently; it does trigger on expressions like `1 << 31` (nicer form would be `1u << 31`) but after clearing a few nuisance ones like that it definitely picked up a few array issues :sweat_smile:

**Pull request merged**